File: Project/Go/project/components/echo/dave_echo.go
package echo
/*
 * Copyright (c) 2023 Renwei
 *
 * This is a free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

import (
	"dave/public/auto"
	"dave/public/base"
	"dave/public/tools"
    "unsafe"
)

var HOW_MANY_CYCLES_DO_STATISTICS = 500
var CONCURRENCY_TPS = 5000

var S8_ECHO_VALUE = -12
var U8_ECHO_VALUE = 12
var S16_ECHO_VALUE = -1234
var U16_ECHO_VALUE = 1234
var S32_ECHO_VALUE = -6462522
var U32_ECHO_VALUE = 35554553
var S64_ECHO_VALUE = -8376462522
var U64_ECHO_VALUE = 83635554553
var FLOAT_ECHO_VALUE = 12.340000
var DOUBLE_ECHO_VALUE = 123.123000
var VOID_ECHO_VALUE = 739848572524
var STRING_ECHO_VALUE = "string echo!"
var MBUF_ECHO_VALUE = "mbuf echo!"

var _echo_working bool = false
var _echo_req_counter uint64 = 0

func _echo_rpc_reset(echo * auto.MsgIdEcho) {
	echo.S8_echo = byte(S8_ECHO_VALUE)
	echo.U8_echo = byte(U8_ECHO_VALUE)
	echo.S16_echo = int16(S16_ECHO_VALUE)
	echo.U16_echo = uint16(U16_ECHO_VALUE)
	echo.S32_echo = int32(S32_ECHO_VALUE)
	echo.U32_echo = uint32(U32_ECHO_VALUE)
	echo.S64_echo = int64(S64_ECHO_VALUE)
	echo.U64_echo = uint64(U64_ECHO_VALUE)
	echo.Float_echo = float32(FLOAT_ECHO_VALUE)
	echo.Double_echo = float64(DOUBLE_ECHO_VALUE)
	echo.Void_echo = unsafe.Pointer(&VOID_ECHO_VALUE)
	tools.T_cgo_gostring2gobyte(echo.String_echo[:], STRING_ECHO_VALUE)
	echo.Mbuf_echo = base.T_gostring2mbuf(MBUF_ECHO_VALUE)
}

func _echo_rpc_copy(echodst * auto.MsgIdEcho, echosrc * auto.MsgIdEcho) {
	echodst.S8_echo = echosrc.S8_echo
	echodst.U8_echo = echosrc.U8_echo
	echodst.S16_echo = echosrc.S16_echo
	echodst.U16_echo = echosrc.U16_echo
	echodst.S32_echo = echosrc.S32_echo
	echodst.U32_echo = echosrc.U32_echo
	echodst.S64_echo = echosrc.S64_echo
	echodst.U64_echo = echosrc.U64_echo
	echodst.Float_echo = echosrc.Float_echo
	echodst.Double_echo = echosrc.Double_echo
	echodst.Void_echo = echosrc.Void_echo
	tools.T_cgo_gostring2gobyte(echodst.String_echo[:], tools.T_cgo_gobyte2gostring(echosrc.String_echo[:]))
	echodst.Mbuf_echo = base.Dave_mclone(echosrc.Mbuf_echo)
}

func _echo_rpc_verification(echo * auto.MsgIdEcho) {
	if echo.S8_echo != byte(S8_ECHO_VALUE) {
		base.DAVELOG("echo.S8_echo != byte(S8_ECHO_VALUE)")
	}
	if echo.U8_echo != byte(U8_ECHO_VALUE) {
		base.DAVELOG("echo.U8_echo != byte(U8_ECHO_VALUE)")
	}
	if echo.S16_echo != int16(S16_ECHO_VALUE) {
		base.DAVELOG("echo.S16_echo != int16(S16_ECHO_VALUE)")
	}
	if echo.U16_echo != uint16(U16_ECHO_VALUE) {
		base.DAVELOG("echo.U16_echo != uint16(U16_ECHO_VALUE)")
	}
	if echo.S32_echo != int32(S32_ECHO_VALUE) {
		base.DAVELOG("echo.S32_echo != int32(S32_ECHO_VALUE)")
	}
	if echo.U32_echo != uint32(U32_ECHO_VALUE) {
		base.DAVELOG("echo.U32_echo != uint32(U32_ECHO_VALUE)")
	}
	if echo.S64_echo != int64(S64_ECHO_VALUE) {
		base.DAVELOG("echo.S64_echo != int64(S64_ECHO_VALUE)")
	}
	if echo.U64_echo != uint64(U64_ECHO_VALUE) {
		base.DAVELOG("echo.U64_echo != uint64(U64_ECHO_VALUE)")
	}
	if echo.Float_echo != float32(FLOAT_ECHO_VALUE) {
		base.DAVELOG("echo.Float_echo != float32(FLOAT_ECHO_VALUE)")
	}
	if echo.Double_echo != float64(DOUBLE_ECHO_VALUE) {
		base.DAVELOG("echo.Double_echo != float64(DOUBLE_ECHO_VALUE)")
	}
	if echo.Void_echo != unsafe.Pointer(&VOID_ECHO_VALUE) {
		base.DAVELOG("echo.Void_echo != unsafe.Pointer(&VOID_ECHO_VALUE)")
	}
	if tools.T_cgo_gobyte2gostring(echo.String_echo[:]) != STRING_ECHO_VALUE {
		base.DAVELOG("tools.T_cgo_gobyte2gostring(echo.String_echo[:]) != STRING_ECHO_VALUE")
	}
	if base.T_mbuf2gostring(echo.Mbuf_echo) != MBUF_ECHO_VALUE {
		base.DAVELOG("tools.T_cgo_gobyte2gostring(echo.Mbuf_echo.Payload[:]) != MBUF_ECHO_VALUE")
	}
}

func _echo_rpc_clean(echo * auto.MsgIdEcho) {
	base.Dave_mfree(echo.Mbuf_echo)
}

func _echo_api_req_msg(gid string, thread string, req auto.MsgIdEchoReq) {
	switch_rand := tools.T_rand_ub() % 4

	if switch_rand == 0 {
		base.Write_msg(thread, auto.MSGID_ECHO_REQ, int(unsafe.Sizeof(req)), unsafe.Pointer(&req))
	} else if switch_rand == 1 {
		base.Write_qmsg(thread, auto.MSGID_ECHO_REQ, int(unsafe.Sizeof(req)), unsafe.Pointer(&req))
	} else if switch_rand == 2 {
		base.Gid_msg(gid, thread, auto.MSGID_ECHO_REQ, int(unsafe.Sizeof(req)), unsafe.Pointer(&req))
	} else if switch_rand == 3 {
		base.Gid_qmsg(gid, thread, auto.MSGID_ECHO_REQ, int(unsafe.Sizeof(req)), unsafe.Pointer(&req))
	}
}

func _echo_api_req_co(gid string, thread string, req auto.MsgIdEchoReq) {
	switch_rand := tools.T_rand_ub() % 4

	pRsp := (*auto.MsgIdEchoRsp)(nil)

	if switch_rand == 0 {
		pRsp = (*auto.MsgIdEchoRsp)(base.Write_co(thread, auto.MSGID_ECHO_REQ, int(unsafe.Sizeof(req)), unsafe.Pointer(&req), auto.MSGID_ECHO_RSP))
	} else if switch_rand == 1 {
		pRsp = (*auto.MsgIdEchoRsp)(base.Write_qco(thread, auto.MSGID_ECHO_REQ, int(unsafe.Sizeof(req)), unsafe.Pointer(&req), auto.MSGID_ECHO_RSP))
	} else if switch_rand == 2 {
		pRsp = (*auto.MsgIdEchoRsp)(base.Gid_co(gid, thread, auto.MSGID_ECHO_REQ, int(unsafe.Sizeof(req)), unsafe.Pointer(&req), auto.MSGID_ECHO_RSP))
	} else if switch_rand == 3 {
		pRsp = (*auto.MsgIdEchoRsp)(base.Gid_qco(gid, thread, auto.MSGID_ECHO_REQ, int(unsafe.Sizeof(req)), unsafe.Pointer(&req), auto.MSGID_ECHO_RSP))
	}

	if pRsp != nil {
		_echo_rpc_clean(&(pRsp.Echo))
	}
}

func _echo_api_req(gid string, thread string, req auto.MsgIdEchoReq) {
	_echo_req_counter += 1

	if ((req.Echo.Type == int32(auto.EchoType_random)) && 
		(_echo_req_counter % 256 == 0) && 
		((tools.T_rand_ub() % 16) == 0)) {
		_echo_api_req_msg(gid, thread, req)
	} else {
		_echo_api_req_msg(gid, thread, req)
	}
}

func _echo_api_rsp(dst uint64, rsp auto.MsgIdEchoRsp) {
	base.Write_msg(dst, auto.MSGID_ECHO_RSP, int(unsafe.Sizeof(rsp)), unsafe.Pointer(&rsp))
}

func _echo_snd_req(gid string, thread string, echo_type int64, getecho auto.MsgIdEcho) {
	req := auto.MsgIdEchoReq{}

	req.Echo = getecho
	_echo_rpc_reset(&(req.Echo))

	req.Echo.Type = int32(echo_type)
	tools.T_cgo_gostring2gobyte(req.Echo.Gid[:], base.Globally_identifier())
	tools.T_cgo_gostring2gobyte(req.Echo.Thread[:], base.Thread_self())

	req.Echo.Echo_req_time = tools.T_time_current_us()
	req.Echo.Echo_rsp_time = 0

	req.Ptr = 0

	_echo_api_req(gid, thread, req)
}

func _echo_snd_rsp(dst uint64, echo_type int64, getecho auto.MsgIdEcho, ptr uint64) {
	rsp := auto.MsgIdEchoRsp{}

	rsp.Echo = getecho
	_echo_rpc_copy(&(rsp.Echo), &getecho)

	rsp.Echo.Type = int32(echo_type)
	tools.T_cgo_gostring2gobyte(rsp.Echo.Gid[:], base.Globally_identifier())
	tools.T_cgo_gostring2gobyte(rsp.Echo.Thread[:], base.Thread_self())

	rsp.Echo.Echo_req_time = getecho.Echo_req_time
	rsp.Echo.Echo_rsp_time = tools.T_time_current_us()

	rsp.Ptr = ptr

	_echo_api_rsp(dst, rsp)
}

func _echo_random(gid string, thread string, getecho auto.MsgIdEcho, random_send_times uint64) {
	for i := uint64(0); i < random_send_times; i++ {
		_echo_snd_req(gid, thread, auto.EchoType_random, getecho)
	}
}

func _echo_concurrent(gid string, thread string, getecho auto.MsgIdEcho) auto.MsgIdEcho {
	if getecho.Concurrent_flag == 1 {
		random_send_times := tools.T_rand_ub() % 128
		if random_send_times == 0 {
			random_send_times = 1
		}

		if (tools.T_time_current_us() - getecho.Concurrent_tps_time) >= 1000000 {
			getecho.Concurrent_tps_time = tools.T_time_current_us()
			getecho.Concurrent_tps_counter = 0
		}

		if getecho.Concurrent_tps_counter < uint64(CONCURRENCY_TPS) {
			if getecho.Concurrent_tps_counter + random_send_times > uint64(CONCURRENCY_TPS) {
				random_send_times = uint64(CONCURRENCY_TPS) - getecho.Concurrent_tps_counter
				if random_send_times >= uint64(CONCURRENCY_TPS) {
					random_send_times = 0
				}
			}

			getecho.Concurrent_tps_counter += random_send_times
			getecho.Concurrent_cycle_counter += random_send_times
			getecho.Concurrent_total_counter += random_send_times

			_echo_random(gid, thread, getecho, random_send_times)
		}
	}

	return getecho
}

func _echo_start(concurrent_flag int8) {
	if _echo_working == true {
		base.DAVELOG("The ECHO system is working!")
		return
	}
	_echo_working = true

	if concurrent_flag == 1 {
		base.DAVELOG("start concurrent echo ...")
	} else {
		base.DAVELOG("start single echo ...")
	}

	req := auto.MsgIdEchoReq{}

	req.Echo.Type = int32(auto.EchoType_single)

	tools.T_cgo_gostring2gobyte(req.Echo.Gid[:], base.Globally_identifier())
	tools.T_cgo_gostring2gobyte(req.Echo.Thread[:], base.Thread_self())

	req.Echo.Echo_total_counter = 0
	req.Echo.Echo_total_time = 0

	req.Echo.Echo_cycle_counter = 0
	req.Echo.Echo_cycle_time = 0

	req.Echo.Echo_req_time = tools.T_time_current_us()
	req.Echo.Echo_rsp_time = 0

	req.Echo.Concurrent_flag = concurrent_flag
	req.Echo.Concurrent_tps_time = 0
	req.Echo.Concurrent_tps_counter = 0
	req.Echo.Concurrent_cycle_counter = 0
	req.Echo.Concurrent_total_counter = 0

	_echo_rpc_reset(&(req.Echo))

	req.Ptr = 0

	base.Broadcast_msg("", auto.MSGID_ECHO_REQ, int(unsafe.Sizeof(req)), unsafe.Pointer(&req))
}

func _echo_stop() {
	if _echo_working == false {
		base.DAVELOG("The ECHO system is not working!")
		return
	}
	_echo_working = false
}

func _echo_single_req(src uint64, getecho auto.MsgIdEcho, ptr uint64) {
	gid := tools.T_cgo_gobyte2gostring(getecho.Gid[:])
	thread := tools.T_cgo_gobyte2gostring(getecho.Thread[:])

	getecho = _echo_concurrent(gid, thread, getecho)

	_echo_snd_rsp(src, auto.EchoType_single, getecho, ptr)
}

func _echo_single_rsp(src uint64, getecho auto.MsgIdEcho) {
	echo_consume_time := tools.T_time_current_us() - getecho.Echo_req_time

	getecho.Echo_total_counter += 1
	getecho.Echo_total_time += echo_consume_time

	getecho.Echo_cycle_counter += 1
	getecho.Echo_cycle_time += echo_consume_time

	if getecho.Echo_cycle_counter >= uint64(HOW_MANY_CYCLES_DO_STATISTICS) {
		base.DAVELOG("%s/%s C:%ds/%d T:%ds/%d %dus/%dus %d",
			tools.T_cgo_gobyte2gostring(getecho.Gid[:]), tools.T_cgo_gobyte2gostring(getecho.Thread[:]),
			getecho.Echo_cycle_time/1000000, getecho.Echo_cycle_counter,
			getecho.Echo_total_time/1000000, getecho.Echo_total_counter,
			getecho.Echo_cycle_time/(getecho.Echo_cycle_counter*2 + getecho.Concurrent_cycle_counter*2),
			getecho.Echo_total_time/(getecho.Echo_total_counter*2 + getecho.Concurrent_total_counter*2),
			getecho.Echo_total_counter+getecho.Concurrent_total_counter)

		getecho.Echo_cycle_counter = 0
		getecho.Concurrent_cycle_counter = 0
		getecho.Echo_cycle_time = 0
	}

	if _echo_working == true {
		gid := tools.T_cgo_gobyte2gostring(getecho.Gid[:])
		thread := tools.T_cgo_gobyte2gostring(getecho.Thread[:])

		getecho = _echo_concurrent(gid, thread, getecho)

		_echo_snd_req(gid, thread, auto.EchoType_single, getecho)
	}
}

func _echo_random_req(src uint64, getecho auto.MsgIdEcho, ptr uint64) {
	_echo_snd_rsp(src, auto.EchoType_random, getecho, ptr)
}

func _echo_random_rsp(src uint64, getecho auto.MsgIdEcho) {
	// Don't do anything.
}

func _echo_req(src uint64, msg_body unsafe.Pointer) {
	pReq := (*auto.MsgIdEchoReq)(msg_body)

	if pReq.Echo.Type == int32(auto.EchoType_start) {
		_echo_start(pReq.Echo.Concurrent_flag)
	} else if pReq.Echo.Type == int32(auto.EchoType_stop) {
		_echo_stop()
	} else if pReq.Echo.Type == int32(auto.EchoType_single) {
		_echo_single_req(src, pReq.Echo, pReq.Ptr)
	} else if pReq.Echo.Type == int32(auto.EchoType_random) {
		_echo_random_req(src, pReq.Echo, pReq.Ptr)
	}

	_echo_rpc_clean(&(pReq.Echo))
}

func _echo_rsp(src uint64, msg_body unsafe.Pointer) {
	pRsp := (*auto.MsgIdEchoRsp)(msg_body)

	if pRsp.Echo.Type == int32(auto.EchoType_single) {
		_echo_rpc_verification(&(pRsp.Echo))

		_echo_single_rsp(src, pRsp.Echo)
	} else if pRsp.Echo.Type == int32(auto.EchoType_random) {
		_echo_random_rsp(src, pRsp.Echo)
	}

	_echo_rpc_clean(&(pRsp.Echo))
}

func _echo(src uint64, msg_id uint64, msg_body unsafe.Pointer) {
	if msg_id == auto.MSGID_ECHO_REQ {
		_echo_req(src, msg_body)
	} else if msg_id == auto.MSGID_ECHO_RSP {
		_echo_rsp(src, msg_body)
	}
} 

func _echo_req_reg(src_gid string, src_name string, src_id uint64, msg_len uint64, msg_body unsafe.Pointer) {
	_echo(src_id, auto.MSGID_ECHO_REQ, msg_body)
}

func _echo_rsp_reg(src_gid string, src_name string, src_id uint64, msg_len uint64, msg_body unsafe.Pointer) {
	_echo(src_id, auto.MSGID_ECHO_RSP, msg_body)
}

// =====================================================================

func Dave_echo_reg() {
	base.Dave_system_function_table_add(auto.MSGID_ECHO_REQ, _echo_req_reg)
	base.Dave_system_function_table_add(auto.MSGID_ECHO_RSP, _echo_rsp_reg)
}

func Dave_echo_unreg() {
	base.Dave_system_function_table_del(auto.MSGID_ECHO_REQ)
	base.Dave_system_function_table_del(auto.MSGID_ECHO_RSP)
}

File: Project/Python/project/components/echo/dave_echo.py
# -*- coding: utf-8 -*-
#/*
# * Copyright (c) 2023 Renwei
# *
# * This is a free software; you can redistribute it and/or modify
# * it under the terms of the MIT license. See LICENSE for details.
# */
from public.base import *
from public.tools import *


HOW_MANY_CYCLES_DO_STATISTICS = 500
CONCURRENCY_TPS = 5000

S8_ECHO_VALUE = b'c'
U8_ECHO_VALUE = b'c'
S16_ECHO_VALUE = -1234
U16_ECHO_VALUE = 1234
S32_ECHO_VALUE = -6462522
U32_ECHO_VALUE = 35554553
S64_ECHO_VALUE = -8376462522
U64_ECHO_VALUE = 83635554553
FLOAT_ECHO_VALUE = 12.340000
DOUBLE_ECHO_VALUE = 123.123000
VOID_ECHO_VALUE = 739848572524
STRING_ECHO_VALUE = "string echo!"
MBUF_ECHO_VALUE = "mbuf echo!"

_echo_working = False
_echo_req_counter = 0


def _echo_rpc_reset(echo):
    echo.s8_echo = S8_ECHO_VALUE
    echo.u8_echo = U8_ECHO_VALUE
    echo.s16_echo = S16_ECHO_VALUE
    echo.u16_echo = U16_ECHO_VALUE
    echo.s32_echo = S32_ECHO_VALUE
    echo.u32_echo = U32_ECHO_VALUE
    echo.s64_echo = S64_ECHO_VALUE
    echo.u64_echo = U64_ECHO_VALUE
    echo.float_echo = FLOAT_ECHO_VALUE
    echo.double_echo = DOUBLE_ECHO_VALUE
    echo.string_echo = bytes(STRING_ECHO_VALUE, encoding='utf8')
    echo.mbuf_echo = str_to_mbuf(MBUF_ECHO_VALUE)
    return echo


def _echo_rpc_copy(echodst, echosrc):
    echodst.s8_echo = echosrc.s8_echo
    echodst.u8_echo = echosrc.u8_echo
    echodst.s16_echo = echosrc.s16_echo
    echodst.u16_echo = echosrc.u16_echo
    echodst.s32_echo = echosrc.s32_echo
    echodst.u32_echo = echosrc.u32_echo
    echodst.s64_echo = echosrc.s64_echo
    echodst.u64_echo = echosrc.u64_echo
    echodst.float_echo = echosrc.float_echo
    echodst.double_echo = echosrc.double_echo
    echodst.void_echo = echosrc.void_echo
    echodst.string_echo = echosrc.string_echo
    echodst.mbuf_echo = dave_mclone(echosrc.mbuf_echo)
    return echodst


def _echo_rpc_verification(echo):
    if echo.s8_echo != S8_ECHO_VALUE:
        DAVELOG(f"echo.s8_echo = {echo.s8_echo} != {S8_ECHO_VALUE}")
    if echo.u8_echo != U8_ECHO_VALUE:
        DAVELOG(f"echo.u8_echo = {echo.u8_echo} != {U8_ECHO_VALUE}")
    if echo.s16_echo != S16_ECHO_VALUE:
        DAVELOG(f"echo.s16_echo = {echo.s16_echo} != {S16_ECHO_VALUE}")
    if echo.u16_echo != U16_ECHO_VALUE:
        DAVELOG(f"echo.u16_echo = {echo.u16_echo} != {U16_ECHO_VALUE}")
    if echo.s32_echo != S32_ECHO_VALUE:
        DAVELOG(f"echo.s32_echo = {echo.s32_echo} != {S32_ECHO_VALUE}")
    if echo.u32_echo != U32_ECHO_VALUE:
        DAVELOG(f"echo.u32_echo = {echo.u32_echo} != {U32_ECHO_VALUE}")
    if echo.s64_echo != S64_ECHO_VALUE:
        DAVELOG(f"echo.s64_echo = {echo.s64_echo} != {S64_ECHO_VALUE}")
    if echo.u64_echo != U64_ECHO_VALUE:
        DAVELOG(f"echo.u64_echo = {echo.u64_echo} != {U64_ECHO_VALUE}")
    if echo.string_echo != bytes(STRING_ECHO_VALUE, encoding='utf8'):
        DAVELOG(f"echo.string_echo = {echo.string_echo} != {STRING_ECHO_VALUE}")
    if mbuf_to_str(echo.mbuf_echo) != (MBUF_ECHO_VALUE):
        DAVELOG(f"echo.mbuf_echo = {mbuf_to_str(echo.mbuf_echo)} != {MBUF_ECHO_VALUE}")
    return


def _echo_rpc_clean(echo):
    dave_mfree(echo.mbuf_echo)
    return


def _echo_api_req_co(gid, thread, pReq):
    switch_rand = t_rand_ub() % 4

    if switch_rand == 0:
        write_co(thread, MSGID_ECHO_REQ, pReq)
    elif switch_rand == 1:
        write_qco(thread, MSGID_ECHO_REQ, pReq)
    elif switch_rand == 2:
        gid_co(gid, thread, MSGID_ECHO_REQ, pReq)
    elif switch_rand == 3:
        gid_qco(gid, thread, MSGID_ECHO_REQ, pReq)
    return


def _echo_api_req_msg(gid, thread, pReq):
    switch_rand = t_rand_ub() % 4

    if switch_rand == 0:
        write_msg(thread, MSGID_ECHO_REQ, pReq)
    elif switch_rand == 1:
        write_qmsg(thread, MSGID_ECHO_REQ, pReq)
    elif switch_rand == 2:
        gid_msg(gid, thread, MSGID_ECHO_REQ, pReq)
    elif switch_rand == 3:
        gid_qmsg(gid, thread, MSGID_ECHO_REQ, pReq)
    return


def _echo_api_req(gid, thread, pReq):
    global _echo_req_counter
    _echo_req_counter += 1

    if (pReq.contents.echo.type == EchoType_random) and \
        (_echo_req_counter % 256 == 0) and \
        ((t_rand_ub() % 16) == 0):
        _echo_api_req_co(gid, thread, pReq)
    else:
        _echo_api_req_msg(gid, thread, pReq)
    return


def _echo_api_rsp(dst, pRsp):
    write_msg(dst, MSGID_ECHO_RSP, pRsp)
    return


def _echo_snd_req(gid, thread, echo_type, getecho):
    pReq = thread_msg(MsgIdEchoReq)

    pReq.contents.echo = getecho
    pReq.contents.echo = _echo_rpc_reset(pReq.contents.echo)

    pReq.contents.echo.type = echo_type
    pReq.contents.echo.gid = bytes(globally_identifier(), encoding='utf8')
    pReq.contents.echo.thread = bytes(thread_self(), encoding='utf8')

    pReq.contents.echo.echo_req_time = t_time_current_us()
    pReq.contents.echo.echo_rsp_time = 0

    pReq.contents.ptr = None

    _echo_api_req(gid, thread, pReq)
    return


def _echo_snd_rsp(dst, echo_type, getecho, ptr):
    pRsp = thread_msg(MsgIdEchoRsp)

    pRsp.contents.echo = getecho
    pRsp.contents.echo = _echo_rpc_copy(pRsp.contents.echo, getecho)

    pRsp.contents.echo.type = echo_type
    pRsp.contents.echo.gid = bytes(globally_identifier(), encoding='utf8')
    pRsp.contents.echo.thread = bytes(thread_self(), encoding='utf8')

    pRsp.contents.echo.echo_req_time = getecho.echo_req_time
    pRsp.contents.echo.echo_rsp_time = t_time_current_us()

    pRsp.contents.ptr = ptr

    _echo_api_rsp(dst, pRsp)
    return


def _echo_random(gid, thread, getecho, random_send_times):
    for i in range(random_send_times):
        _echo_snd_req(gid, thread, EchoType_random, getecho)
    return


def _echo_concurrent(gid, thread, getecho):
    if getecho.concurrent_flag == True:
        random_send_times = t_rand_ub() % 128
        if random_send_times == 0:
            random_send_times = 1

        if t_time_current_us() - getecho.concurrent_tps_time > 1000000:
            getecho.concurrent_tps_time = t_time_current_us()
            getecho.concurrent_tps_counter = 0
        
        if getecho.concurrent_tps_counter < CONCURRENCY_TPS:
            if getecho.cuncurrent_tps_counter + random_send_times > CONCURRENCY_TPS:
                random_send_times = CONCURRENCY_TPS - getecho.concurrent_tps_counter
                if random_send_times >= CONCURRENCY_TPS:
                    random_send_times = 0

                getecho.concurrent_tps_counter += random_send_times
                getecho.concurrent_cycle_counter += random_send_times
                getecho.concurrent_total_counter += random_send_times

                _echo_random(gid, thread, getecho, random_send_times)
    return getecho


def _echo_start(concurrent_flag):
    global _echo_working
    if _echo_working == True:
        DAVELOG("The ECHO system is working!")
        return
    _echo_working = True

    if concurrent_flag:
        DAVELOG("start concurrent echo ...")
    else:
        DAVELOG("start single echo ...")

    pReq = thread_msg(MsgIdEchoReq)

    pReq.contents.echo.type = EchoType_single

    pReq.contents.echo.gid = bytes(globally_identifier(), encoding='utf8')
    pReq.contents.echo.thread = bytes(thread_self(), encoding='utf8')

    pReq.contents.echo.echo_total_counter = 0
    pReq.contents.echo.echo_total_time = 0

    pReq.contents.echo.echo_cycle_counter = 0
    pReq.contents.echo.echo_cycle_time = 0
    pReq.contents.echo.echo_req_time = t_time_current_us()
    pReq.contents.echo.echo_rsp_time = 0

    pReq.contents.echo.concurrent_flag = concurrent_flag
    pReq.contents.echo.concurrent_tps_time = 0
    pReq.contents.echo.concurrent_tps_counter = 0
    pReq.contents.echo.concurrent_cycle_counter = 0
    pReq.contents.echo.concurrent_total_counter = 0

    pReq.echo = _echo_rpc_reset(pReq.contents.echo)

    pReq.contents.ptr = None

    broadcast_msg("", MSGID_ECHO_REQ, pReq)
    return


def _echo_stop():
    global _echo_working
    if _echo_working == False:
        DAVELOG("The ECHO system is not working!")
        return
    _echo_working = False
    return


def _echo_single_req(src, getecho, ptr):
    getecho = _echo_concurrent(getecho.gid, getecho.thread, getecho)

    _echo_snd_rsp(src, EchoType_single, getecho, ptr)
    return


def _echo_single_rsp(src, getecho):
    echo_consume_time = t_time_current_us() - getecho.echo_req_time

    getecho.echo_total_counter += 1
    getecho.echo_total_time += echo_consume_time

    getecho.echo_cycle_counter += 1
    getecho.echo_cycle_time += echo_consume_time

    if getecho.echo_cycle_counter >= HOW_MANY_CYCLES_DO_STATISTICS:
        DAVELOG(f"{getecho.gid}/{getecho.thread} "\
                f"C:{int(getecho.echo_cycle_time/1000000)}s/{getecho.echo_cycle_counter} "\
                f"T:{int(getecho.echo_total_time/1000000)}s/{getecho.echo_total_counter} "\
                f"{int(getecho.echo_cycle_time/(getecho.echo_cycle_counter*2 + getecho.concurrent_cycle_counter*2))}us/{int(getecho.echo_total_time/(getecho.echo_total_counter*2 + getecho.concurrent_total_counter*2))}us "\
                f"{getecho.echo_total_counter+getecho.concurrent_total_counter}")

        getecho.echo_cycle_counter = 0
        getecho.concurrent_cycle_counter = 0
        getecho.echo_cycle_time = 0

    global _echo_working
    if _echo_working == True:
        getecho = _echo_concurrent(getecho.gid, getecho.thread, getecho)

        _echo_snd_req(getecho.gid, getecho.thread, EchoType_single, getecho)

    return


def _echo_random_req(src, getecho, ptr):
    _echo_snd_rsp(src, EchoType_random, getecho, ptr)
    return


def _echo_random_rsp(src, getecho):
    # Don't do anything.
    return


def _echo_req(src, msg_len, msg_body):
    pReq = struct_copy(MsgIdEchoReq, msg_body, msg_len)

    if pReq.echo.type == EchoType_start:
        _echo_start(pReq.echo.concurrent_flag)
    elif pReq.echo.type == EchoType_stop:
        _echo_stop()
    elif pReq.echo.type == EchoType_single:
        _echo_single_req(src, pReq.echo, pReq.ptr)
    elif pReq.echo.type == EchoType_random:
        _echo_random_req(src, pReq.echo, pReq.ptr)
		
    _echo_rpc_clean(pReq.echo)
    return


def _echo_rsp(src, msg_len, msg_body):
    pRsp = struct_copy(MsgIdEchoRsp, msg_body, msg_len)

    if pRsp.echo.type == EchoType_single:
        _echo_rpc_verification(pRsp.echo)

        _echo_single_rsp(src, pRsp.echo)
    elif pRsp.echo.type == EchoType_random:
        _echo_random_rsp(src, pRsp.echo)
		
    _echo_rpc_clean(pRsp.echo)
    return


def _echo(src, msg_id, msg_len, msg_body):
    if msg_id == MSGID_ECHO_REQ:
        _echo_req(src, msg_len, msg_body)
    elif msg_id == MSGID_ECHO_RSP:
        _echo_rsp(src, msg_len, msg_body)
    return


def _echo_req_reg(src_name, src_id, msg_len, msg_body):
    _echo(src_id, MSGID_ECHO_REQ, msg_len, msg_body)
    return


def _echo_rsp_reg(src_name, src_id, msg_len, msg_body):
    _echo(src_id, MSGID_ECHO_RSP, msg_len, msg_body)
    return


# =====================================================================


def dave_echo_reg():
    dave_system_function_table_add(MSGID_ECHO_REQ, _echo_req_reg)
    dave_system_function_table_add(MSGID_ECHO_RSP, _echo_rsp_reg)
    return

def dave_echo_unreg():
    dave_system_function_table_del(MSGID_ECHO_REQ)
    dave_system_function_table_del(MSGID_ECHO_RSP)
    return